### ISSUE TITLE
Enumerator.peek not working with Enumerators created with #to_enum over ActiveRecord Relations

### ISSUE THREAD

**Issue by dgolombek** [jruby/jruby#6309]
**Environment Information**
  
* JRuby version: jruby 9.2.11.1 (2.5.7) 2020-03-25 b1f55b1a40 OpenJDK 64-Bit Server VM 25.232-b09 on 1.8.0_232-b09 +jit [darwin-x86_64]
* OS: Darwin MBP.local 18.7.0 Darwin Kernel Version 18.7.0: Mon Feb 10 21:08:45 PST 2020; root:xnu-4903.278.28~1/RELEASE_X86_64 x86_64
* ActiveRecord *: 4.2.11.3

**Expected Behavior**

When I create an Enumerator over an ActiveRecord Relation using #to_enum, #peek on that enumerator does not always work. The exact behavior seems to vary a bit -- in my sample test, #peek throws StopIteration, but when I run this in larger environments, the Enumerator seems to hang. This may be tied to to AR Model size (and thus load time) -- I have a harder time reproducing with my small models, and easier with big, bloated objects. 

https://gist.github.com/dgolombek/3123d5602ff6d76e96f2bac01b570210

**Actual Behavior**

> Failures:
>
>  1) Enumerator with a single value in an AR array should not hang
>     Failure/Error: raise "bad first element on first peek, #{iter.peek.id} != #{first.id}" unless iter.peek.id == first.id
>
>     StopIteration:
>       iteration reached an end
>     # uri:classloader:/jruby/kernel/enumerator.rb:25:in `next'
>     # uri:classloader:/jruby/kernel/enumerator.rb:34:in `peek'
>     # ./spec/simple_enum_spec.rb:4:in `check_iter'
>     # ./spec/simple_enum_spec.rb:25:in `block in <main>'
>     # ./spec/spec_helper.rb:113:in `block in <main>'
>     # ./spec/spec_helper.rb:112:in `block in <main>'
>     # ./spec/spec_helper.rb:111:in `block in <main>'
>     # ./spec/spec_helper.rb:100:in `block in <main>'
>     # ./spec/spec_helper.rb:99:in `block in <main>'

Using MRI 2.6.6, this works correctly. As mentioned above, when I deployed to prod, we saw uncontrolled thread growth, leading to server death, which I believe was tied to threads hanging in the enumerator (but could be tied to bugs with our handling of unexpected StopIteration, I'm still debugging that). I haven't gotten thread dumps yet, will work on doing so.

**Comment by enebo** [comment]
@dgolombek  Can you supply your factory bot definition for :device?

**Comment by dgolombek** [comment]
```
FactoryBot.define do
  factory :device, class: Models::Device do |d|
    d.association       :tenant, factory: :tenant_without_mdm_connection, strategy: :create
    d.guid              { SecureRandom.uuid }
    d.tenant_device_id  { SecureRandom.uuid }
    d.state             { Models::Device.states[:pending] } 
    d.platform          { Models::Device::UNKNOWN_PLATFORM }
    d.is_invalid        { false }
  end
end
```

The DB table has 3 int(11), 5 varchar(255), 5 tinyint(1), and 2 datetimes in it.

**Comment by dgolombek** [comment]
Fails the same way in 9.2.12.0

**Comment by enebo** [comment]
Getting closer but I am hoping for enough in this spec to be able to run it.  Models::Device is the next missing piece.  It is likely once we run this we will realize it can be reproduced to something much smaller but being able to reproduce this will help us get to that point.

**Comment by dgolombek** [comment]
I can reproduce using this model:

```
module Models
  class SimpleDevice < ActiveRecord::Base
    self.table_name = "devices"

    belongs_to :tenant
  end
end
```

The full SQL:

```
DROP TABLE IF EXISTS `devices`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!40101 SET character_set_client = utf8 */;
CREATE TABLE `devices` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `tenant_id` int(11) unsigned NOT NULL,
  `tenant_device_id` varchar(255) NOT NULL,
  `guid` varchar(255) NOT NULL,
  `state` tinyint(1) unsigned NOT NULL DEFAULT '3',
  `state_sync_needed` tinyint(1) DEFAULT '0',
  `details_hash` varchar(255) DEFAULT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `unsupported_reason` varchar(255) DEFAULT NULL,
  `is_invalid` tinyint(1) NOT NULL DEFAULT '0',
  `platform` int(11) NOT NULL DEFAULT '3',
  `inventory_enabled` tinyint(1) NOT NULL DEFAULT '0',
  `orphan` tinyint(1) NOT NULL DEFAULT '0',
  `related_guid` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `unique_devices_tenants` (`tenant_id`,`tenant_device_id`),
  CONSTRAINT `devices_tenant_fk` FOREIGN KEY (`tenant_id`) REFERENCES `tenants` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
/*!40101 SET character_set_client = @saved_cs_client */;

```

**Comment by headius** [comment]
A full JVM trace might be helpful... get it to hang again and then get a JVM stack trace dump one of these ways:

* Ctrl+`\` in the same terminal (sends signal QUIT to the process)
* Use the `jstack` command on the pid

Also to clarify something in your spec... you use a `.each` enumerator and that works ok? Only the `.to_enum` iterator raises the `StopIteration` error?

The `StopIteration` is not really an error, and would be expected if you iterate an Enumerator past the last element.

We could also get a "full" trace of these exception stacks by having you pass `-Xbacktrace.style=full` to JRuby when you start it up. That will give a bit more information.

**Comment by dgolombek** [comment]
I'll try to reproduce and get a stack trace. It's been tricky, only happens under high load in a way that I've been unable to reproduce with our local stress testing framework.

About the clarification, yes, .each enumerator works correctly and .to_enum iterator fails on .peek when there is an element still in the source. 

Unfortunately, -Xbacktrace.style=full doesn't add much:

```
     Failure/Error: raise "bad first element on first peek, #{iter.peek.id} != #{first.id}" unless iter.peek.id == first.id

     StopIteration:
       iteration reached an end
     # java/lang/Thread.java:1559:in `getStackTrace'
     # uri:classloader:/jruby/kernel/enumerator.rb:25:in `next'
     # uri:classloader:/jruby/kernel/enumerator.rb:34:in `peek'
     # ./spec/helpers/simple_enum_spec.rb:4:in `check_iter'
     # ./spec/helpers/simple_enum_spec.rb:25:in `block in <main>'
     # ./spec/spec_helper.rb:113:in `block in <main>'
     # ./spec/spec_helper.rb:112:in `block in <main>'
     # ./spec/spec_helper.rb:111:in `block in <main>'
     # ./spec/spec_helper.rb:100:in `block in <main>'
     # ./spec/spec_helper.rb:99:in `block in <main>'
     # java/lang/invoke/MethodHandle.java:627:in `invokeWithArguments'

```

**Comment by headius** [comment]
@dgolombek Ok thanks for the clarification.

The fact that it only happens under load makes me wonder if this is either a concurrency issue (other than the fact that fibers are threads in JRuby) or an optimization issue (some code eventually gets optimized incorrectly). The fact that `to_enum` fails and `each` works is also very interesting. Will need to compare those two paths.

**Comment by dgolombek** [comment]
This is now happening for us again with 9.2.14.0 See attached thread dump:

[i07f5_jstack.log](https://github.com/jruby/jruby/files/5850457/i07f5_jstack.log)

**Comment by dgolombek** [comment]
I should note that the collections from the most recent stack dump are created with #each, making this slightly different than the original bug. So perhaps this should be a new issue.

**Comment by dgolombek** [comment]
Adding a new thread dump to show how things evolved over time -- some of the fibers are moving a little bit, rather than being entirely blocked, but the pattern remains the same:
[i07f5_jstack.log](https://github.com/jruby/jruby/files/5851379/i07f5_jstack.log)

**Comment by headius** [comment]
Sounding good according to @dgolombek based on deployment of the patch in #6538 to the app they were having issues with. Fingers crossed... 🤞

**Comment by headius** [comment]
Summarizing what the problem appears to have been:

In the logic for fiber resuming and yielding, we use an interruptible queue push/pop to post values and await results from one thread to the target fiber thread. While waiting for these hand-offs to occur it is possible for someone to interrupt the thread, such as to shut down a system or cancel a runaway thread or fiber. If a thread A is waiting on a fiber B and A receives such an interrupting exception, we want to forward that exception to B since we should technically be operating on B's call stack. We allow B to handle the interrupt and wait for a result again. However if B is near the logic to transfer values back to A, we may end up with a situation where they both think the other should be executing and endlessly forward the exception back and forth.

In order to deal with this situation, we give the target fiber one chance to handle any interrupt exception. We wait once more for a result. If the fiber handles the exception and returns a normal result, we proceed normally. If, however, the fiber tosses the exception back or we are again interrupted, we give up on the fiber and propagate the error.

See the patches in #6538.
